File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/cli/verbs/lint.mjs
import { readFileSync, statSync, readdirSync } from "node:fs";
import { join, basename, resolve } from "node:path";
import {
  parseFrontmatter,
  MissingFrontmatterError,
  MalformedFrontmatterError,
} from "../../parser.mjs";
import { lint } from "../../lint/index.mjs";

function discoverSkills(target) {
  const st = statSync(target);
  if (st.isFile()) {
    if (basename(target) === "SKILL.md") return [resolve(target, "..")];
    return [];
  }
  const direct = join(target, "SKILL.md");
  try {
    if (statSync(direct).isFile()) return [target];
  } catch {
    /* fall through */
  }
  const out = [];
  for (const entry of readdirSync(target, { withFileTypes: true })) {
    if (!entry.isDirectory()) continue;
    if (entry.name.startsWith(".")) continue;
    out.push(...discoverSkills(join(target, entry.name)));
  }
  return out;
}

function parseArgs(args) {
  const opts = {
    path: null,
    format: "human",
    explain: false,
    minScore: 40,
  };
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === "--format") opts.format = args[++i];
    else if (a === "--explain") opts.explain = true;
    else if (a === "--min-score") {
      const v = args[++i];
      const n = parseInt(v, 10);
      if (!Number.isNaN(n)) opts.minScore = n;
    } else if (!a.startsWith("--") && opts.path == null) opts.path = a;
  }
  return opts;
}

function sortKeys(v) {
  if (Array.isArray(v)) return v.map(sortKeys);
  if (v && typeof v === "object") {
    const o = {};
    for (const k of Object.keys(v).sort()) o[k] = sortKeys(v[k]);
    return o;
  }
  return v;
}

function lintOne(skillDir) {
  const skillPath = join(skillDir, "SKILL.md");
  let text;
  try {
    text = readFileSync(skillPath, "utf8");
  } catch (e) {
    return {
      path: skillPath,
      score: 0,
      breakdown: {
        keywordDensity: 0,
        actionVerbs: 0,
        triggerPhrases: 0,
        specificity: 0,
        length: 0,
      },
      issues: [
        {
          code: "E100_SKILL_MD_MISSING",
          severity: "error",
          field: "skill",
          message: `SKILL.md not readable: ${e.message}`,
        },
      ],
    };
  }
  let parsed;
  try {
    parsed = parseFrontmatter(text);
  } catch (e) {
    const code =
      e instanceof MissingFrontmatterError
        ? "E101_FRONTMATTER_MISSING"
        : e instanceof MalformedFrontmatterError
          ? "E102_FRONTMATTER_MALFORMED"
          : "E102_FRONTMATTER_MALFORMED";
    return {
      path: skillPath,
      score: 0,
      breakdown: {
        keywordDensity: 0,
        actionVerbs: 0,
        triggerPhrases: 0,
        specificity: 0,
        length: 0,
      },
      issues: [
        { code, severity: "error", field: "frontmatter", message: e.message },
      ],
    };
  }
  const r = lint(parsed, skillDir);
  return { path: skillPath, ...r };
}

function exitCodeFor(report, minScore) {
  if (report.issues.some((i) => i.severity === "error")) return 1;
  if (report.issues.some((i) => i.code === "W002_BROKEN_REFERENCE")) return 1;
  if (report.score < minScore) return 1;
  return 0;
}

function formatHuman(report, opts) {
  const lines = [];
  const status = exitCodeFor(report, opts.minScore) === 0 ? "OK" : "FAIL";
  lines.push(`${report.path}`);
  lines.push(`  discoverability: ${report.score}/100 [${status}]`);
  if (opts.explain) {
    const b = report.breakdown;
    lines.push(
      `  sub-scores: keyword=${b.keywordDensity}/25 verbs=${b.actionVerbs}/20 trigger=${b.triggerPhrases}/20 specificity=${b.specificity}/20 length=${b.length}/15`,
    );
  }
  for (const i of report.issues) {
    lines.push(`  [${i.severity}] ${i.code}: ${i.message}`);
  }
  return lines.join("\n") + "\n";
}

export async function lintVerb(args) {
  const opts = parseArgs(args);
  if (!opts.path) {
    return {
      exitCode: 64,
      output:
        "usage: skills-ref lint <path> [--format json|human] [--explain] [--min-score N]\n",
    };
  }
  const skills = discoverSkills(resolve(opts.path));
  if (skills.length === 0) {
    return { exitCode: 1, output: `no SKILL.md found under: ${opts.path}\n` };
  }
  const reports = skills.map(lintOne);
  let exitCode = 0;
  for (const r of reports) {
    if (exitCodeFor(r, opts.minScore) === 1) exitCode = 1;
  }
  let output;
  if (opts.format === "json") {
    const payload =
      reports.length === 1
        ? sortKeys({
            score: reports[0].score,
            breakdown: reports[0].breakdown,
            issues: reports[0].issues,
            path: reports[0].path,
          })
        : sortKeys({
            reports: reports.map((r) => ({
              path: r.path,
              score: r.score,
              breakdown: r.breakdown,
              issues: r.issues,
            })),
          });
    output = JSON.stringify(payload) + "\n";
  } else {
    output = reports.map((r) => formatHuman(r, opts)).join("");
  }
  return { exitCode, output };
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/verbs/lint.py
from __future__ import annotations

import json
from pathlib import Path

from ..lint import lint as _lint
from ..parser import (
    MalformedFrontmatterError,
    MissingFrontmatterError,
    parse_frontmatter,
)


_EMPTY_BREAKDOWN = {
    "keywordDensity": 0,
    "actionVerbs": 0,
    "triggerPhrases": 0,
    "specificity": 0,
    "length": 0,
}


def _discover_skills(target: Path) -> list[Path]:
    if target.is_file():
        if target.name == "SKILL.md":
            return [target.parent]
        return []
    direct = target / "SKILL.md"
    if direct.is_file():
        return [target]
    out: list[Path] = []
    for entry in sorted(target.iterdir(), key=lambda p: p.name):
        if not entry.is_dir():
            continue
        if entry.name.startswith("."):
            continue
        out.extend(_discover_skills(entry))
    return out


def _sort(v):
    if isinstance(v, list):
        return [_sort(x) for x in v]
    if isinstance(v, dict):
        return {k: _sort(v[k]) for k in sorted(v.keys())}
    return v


def _lint_one(skill_dir: Path) -> dict:
    skill_path = skill_dir / "SKILL.md"
    try:
        text = skill_path.read_text(encoding="utf-8")
    except OSError as e:
        return {
            "path": str(skill_path),
            "score": 0,
            "breakdown": dict(_EMPTY_BREAKDOWN),
            "issues": [
                {
                    "code": "E100_SKILL_MD_MISSING",
                    "severity": "error",
                    "field": "skill",
                    "message": f"SKILL.md not readable: {e}",
                }
            ],
        }
    try:
        parsed = parse_frontmatter(text)
    except MissingFrontmatterError as e:
        return {
            "path": str(skill_path),
            "score": 0,
            "breakdown": dict(_EMPTY_BREAKDOWN),
            "issues": [
                {
                    "code": "E101_FRONTMATTER_MISSING",
                    "severity": "error",
                    "field": "frontmatter",
                    "message": str(e),
                }
            ],
        }
    except MalformedFrontmatterError as e:
        return {
            "path": str(skill_path),
            "score": 0,
            "breakdown": dict(_EMPTY_BREAKDOWN),
            "issues": [
                {
                    "code": "E102_FRONTMATTER_MALFORMED",
                    "severity": "error",
                    "field": "frontmatter",
                    "message": str(e),
                }
            ],
        }
    r = _lint(parsed, str(skill_dir))
    return {
        "path": str(skill_path),
        "score": r["score"],
        "breakdown": r["breakdown"],
        "issues": r["issues"],
    }


def _exit_code_for(report: dict, min_score: int) -> int:
    if any(i.get("severity") == "error" for i in report["issues"]):
        return 1
    if any(i.get("code") == "W002_BROKEN_REFERENCE" for i in report["issues"]):
        return 1
    if report["score"] < min_score:
        return 1
    return 0


def _format_human(report: dict, explain: bool, min_score: int) -> str:
    lines: list[str] = []
    status = "OK" if _exit_code_for(report, min_score) == 0 else "FAIL"
    lines.append(report["path"])
    lines.append(f"  discoverability: {report['score']}/100 [{status}]")
    if explain:
        b = report["breakdown"]
        lines.append(
            "  sub-scores: "
            f"keyword={b['keywordDensity']}/25 "
            f"verbs={b['actionVerbs']}/20 "
            f"trigger={b['triggerPhrases']}/20 "
            f"specificity={b['specificity']}/20 "
            f"length={b['length']}/15"
        )
    for i in report["issues"]:
        lines.append(f"  [{i['severity']}] {i['code']}: {i['message']}")
    return "\n".join(lines) + "\n"


def lint(args) -> tuple[int, str]:
    path = Path(args.path).resolve()
    if not path.exists():
        return 1, f"no SKILL.md found under: {args.path}\n"
    skills = _discover_skills(path)
    if not skills:
        return 1, f"no SKILL.md found under: {args.path}\n"
    min_score = int(getattr(args, "min_score", 40) or 40)
    explain = bool(getattr(args, "explain", False))
    reports = [_lint_one(d) for d in skills]
    exit_code = 0
    for r in reports:
        if _exit_code_for(r, min_score) == 1:
            exit_code = 1
    if args.format == "json":
        if len(reports) == 1:
            payload = _sort(
                {
                    "score": reports[0]["score"],
                    "breakdown": reports[0]["breakdown"],
                    "issues": reports[0]["issues"],
                    "path": reports[0]["path"],
                }
            )
        else:
            payload = _sort(
                {
                    "reports": [
                        {
                            "path": r["path"],
                            "score": r["score"],
                            "breakdown": r["breakdown"],
                            "issues": r["issues"],
                        }
                        for r in reports
                    ]
                }
            )
        return exit_code, json.dumps(payload, separators=(",", ":")) + "\n"
    out = "".join(_format_human(r, explain, min_score) for r in reports)
    return exit_code, out
